### ISSUE TITLE
`std::io::Chain` could implement BufRead

### ISSUE THREAD

**Issue by troplin** [rust-lang/rust#32536]
The implementation is straight forward:
```rust
impl<T: BufRead, U: BufRead> BufRead for Chain<T, U> {
    fn fill_buf(&mut self) -> Result<&[u8]> {
        if !self.done_first {
            match try!(self.first.fill_buf()) {
                buf if buf.len() == 0 => { self.done_first = true; }
                buf => return Ok(buf),
            }
        }
        self.second.fill_buf()
    }
    fn consume(&mut self, amt: usize) {
        if !self.done_first {
            self.first.consume(amt)
        } else {
            self.second.consume(amt)
        }
    }
}
```
Or am I overlooking something?

**Comment by rphmeier** [comment]
`std::net::Chain` -> `std::io::Chain`?

Now that the `?` operator has landed, I would guess that would be favored over `try!` in `std`.
The tricky thing about trait implementations is that they become insta-stable, so it's usually good to be apprehensive of adding new ones. This one seems uncontroversial, so it's likely that it may have just been overlooked previously.

**Comment by troplin** [comment]
Whoops, yes.
I've changed the title.

The implementation is not set in stone of course.

**Comment by sfackler** [comment]
This seems reasonable I think - want to open a PR?

**Comment by troplin** [comment]
I'll give it a try. I'm still learning how to everything works with github...

**Comment by troplin** [comment]
@rphmeier you mentioned that it will be insta-stable. Does that mean that I have to tag it as stable for my PR?
``` rust
#[stable(feature = "rust1", since = "1.9.0")]
```
Like that?

**Comment by sfackler** [comment]
Yep, though you can't use the `rust1` feature since feature names are tied to a specific release. Just make one up though - `chain_buf_read` or whatever.

**Comment by Aatch** [comment]
Closed by #32541
